pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -375,323 +475,401 @@ func (sf *IteratorDatastore) Read(
+		ctx:      ctx,
+		cancel:   cancel,
+		cleanup:  cleanup,
+		it:       it,
+		fetching: &fetching,
+		items:    &items,
+		err:      &err,
+		refs:     &refs,
+		wg:       &wg,
+		ch:       &pch,
+	}
+
+	return &newIter
 }
 
-// It is assumed that mu is held by the parent while cloning.
-func (s *sharedIterator) clone() (*sharedIterator, error) {
-	if s.initializationErr != nil {
-		return nil, s.initializationErr
-	}
-	if time.Now().After(s.maxAdmissionTime) {
-		// To avoid stale data, we want to prevent shared iterator from using the clone if it is created after
-		// maxAdmissionTime. When we return, the clone will default to skip the shared iterator.
-		return nil, errSharedIteratorAfterLastAdmissionTime
-	}
-
-	newIter := &sharedIterator{
-		manager:      s.manager,
-		key:          s.key,
-		head:         0,
-		maxAliveTime: s.maxAliveTime,
-		// the start time and the maxAdmissionTime is only used for checking
-		// whether we need to clone. As such, it is not copied into the cloned copy.
-
-		mu:                 s.mu,
-		items:              s.items,
-		inner:              s.inner,
-		sharedErr:          s.sharedErr,
-		watchdogTimer:      s.watchdogTimer,
-		watchdogTimeoutErr: s.watchdogTimeoutErr,
-	}
-	newIter.watchdogTimer.Reset(s.maxAliveTime)
-	return newIter, nil
+// clone creates a new shared iterator that shares the same context, cancellation function, and other fields.
+// It increments the reference count and returns a new instance of the shared iterator.
+// If the reference count reaches zero, it returns nil, indicating that the iterator has been stopped.
+// This allows multiple goroutines to share the same iterator instance without interfering with each other.
+// The clone method is thread-safe and ensures that the reference count is incremented atomically.
+func (s *sharedIterator) clone() *sharedIterator {
+	for {
+		remaining := s.refs.Load()
+
+		// If the reference count is zero, it means that the iterator has been stopped and cleaned up.
+		if remaining == 0 {
+			return nil
+		}
+
+		if s.refs.CompareAndSwap(remaining, remaining+1) {
+			return &sharedIterator{
+				ctx:      s.ctx,
+				cancel:   s.cancel,
+				cleanup:  s.cleanup,
+				it:       s.it,
+				fetching: s.fetching,
+				items:    s.items,
+				err:      s.err,
+				refs:     s.refs,
+				wg:       s.wg,
+				ch:       s.ch,
+			}
+		}
+	}
 }
 
-func (s *sharedIterator) cleanup() {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	if s.inner != nil {
-		s.inner.Stop()
+// fetch is a method that fetches items from the underlying storage.TupleIterator.
+// It reads a fixed number of items (BufferSize) from the iterator and stores them in the shared items slice.
+// If there is an error while fetching items, it stores the error in the shared error pointer.
+// It also manages the channel that signals when new items are available.
+// The fetch method is called by the shared iterator instances when they need to fetch new items.
+func (s *sharedIterator) fetch() {
```

Comment:
It may worth calling out that there is only at most a single instance of fetch() running which is protected by s.fetching.  If there are multiple instance of fetch() running per sharedIterator, this will not work,